core.d.ts

```diff
@@ -0,0 +1,145 @@
+interface Util {
+	isString(v?: any): v is string
+	isNumber(v: any): v is number
+	isBoolean(v: any): v is boolean
+	isObject(v: any): v is object
+	isArray(v: any): v is Array<any>
+	isDate(v: any): v is Date
+}
+interface PromiseRejector {
+	reject(err: any): Promise<any>
+}
+
+interface ResourceObj<T> {
+	[index: string]: T
+}
+
+type FilterOperationKey = '$ne' | '$eq' | '$gt' | '$ge' | '$lt' | '$le' | '$add' | '$sub' | '$mul' | '$div' | '$mod'
+type FilterOperationValue = Filter
+type FilterOperations = {
+	[P in FilterOperationKey]: FilterOperationValue
+}
+type FilterFunctionKey = '$contains' | '$endswith' | '$startswith' | '$length' | '$indexof' | '$substring' | '$tolower' | '$toupper' | '$trim' | '$concat' | '$year' | '$month' | '$day' | '$hour' | '$minute' | '$second' | '$fractionalseconds' | '$date' | '$time' | '$totaloffsetminutes' | '$now' | '$maxdatetime' | '$mindatetime' | '$totalseconds' | '$round' | '$floor' | '$ceiling' | '$isof' | '$cast'
+type FilterFunctionValue = Filter
+type FilterFunctions = {
+	[P in FilterFunctionKey]: FilterFunctionValue
+}
+
+interface ResourceFilter extends ResourceObj<Filter> {}
+
+type FilterObj = FilterOperations & FilterFunctions & {
+	$raw?: RawFilter
+
+	$?: string | string[]
+
+	$and?: Filter
+	$or?: Filter
+
+	$in?: Filter
+
+	$not?: Filter
+
+	$any?: Lambda
+	$all?: Lambda
+} & ResourceFilter
+
+interface FilterArray extends Array<Filter> {}
+type FilterString = string
+// Strictly speaking `[ string, ...Filter ]` but there isn't a way to type that yet
+type RawFilter = string | (string | Filter)[] | {
+	$string: string
+	[index: string]: Filter
+}
+type Lambda = {
+	$alias: string
+	$expr: Filter
+}
+type Filter = number | FilterString | FilterObj | FilterArray
+
+interface ResourceExpand extends ResourceObj<Expand> {}
+type ExpandObject = ODataOptions & ResourceExpand
+
+interface ExpandArray extends Array<Expand> {}
+type Expand = string | ExpandArray | ExpandObject
+
+type ODataOptions = {
+	$filter?: Filter
+	$expand?: Expand
+	$orderby?: any
+	$top?: any
+	$skip?: any
```

Comment:
Yeah, I haven't actually fully typed everything yet, I wanted to get this in review though because of other issues where you've already saved me loads of time :D

I've changed these and also $select as string | string[]